inline-r/src/Language/R/Parser.hs

```diff
@@ -0,0 +1,327 @@
+-- |
+-- Copyright:   (c) 2016, AlphaSheets, Inc
+-- Stability:   Experimental
+-- Portability: Portable
+--
+-- "Parser" module provides high-level interface that allows to parse
+-- SEXP values into the complex Haskell data types.
+-- This parser may be useful for application specific solutions
+-- that need to convert different R types to the application specific ADT.
+--
+-- In addition this is the the simplest way to get additional attributes
+-- that exists in the structure.
+--
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE DeriveGeneric #-}
+module Language.R.Parser
+  ( Parser(..)
+  , parseOnly
+    -- * Parser interface.
+    -- $interface
+  , somesexp
+  , sexp
+  , with
+    -- * Type guards
+    -- $guards
+  , hexp
+  , null
+  , s4
+  , s3
+  , guardType
+    -- * Queries
+  , typeOf
+  , getS3Class
+    -- * Attributes
+    -- $attributes
+  , someAttribute
+  , attribute
+  , attributes
+    -- * Attribute parsers.
+  , names
+  , dim
+  , dimnames
+  , rownames
+    -- * Predefined parsers
+    -- $parsers
+  , factor
+    -- * Helpers
+  , charList
+  , choice
+  , list
+  ) where
+
+import Data.Singletons
+import qualified Foreign.R as R
+
+import           Control.DeepSeq
+import           Control.Applicative
+import           Control.Monad (guard, ap, liftM)
+import           Control.Exception (evaluate)
+import           Foreign hiding (void, with)
+import           Foreign.C.String
+import           H.Prelude hiding (typeOf, hexp)
+import qualified H.Prelude as H
+
+import           Data.Typeable (Typeable)
+import           Data.Functor (void)
+import           Data.Foldable (asum)
+import           Data.Traversable
+import           Data.Maybe (mapMaybe)
+import qualified Data.Vector.SEXP as SV
+import           System.IO.Unsafe
+import           GHC.Generics (Generic)
+
+import Prelude hiding (null)
+
+-- | Parser. That run inspection over 'SomeSEXP'.
+-- Parser is bound to the region where 'SomeSEXP' is allocated,
+-- so extracted value will not leak out of the region scope.
+--
+-- This parser is a pure function, so if you need to allocate
+-- any object (for example for comparison or lookup) you should
+-- do it before running parser.
+newtype Parser s r a = Parser {runParser :: SomeSEXP s -> (a -> r) -> (ParserError s -> r) -> r}
```

Comment:
Document the meaning of `r`.

Document the meaning of the arguments of `runParser`.